### PR TITLE
Digital Twin pre-release candidate

### PR DESCRIPTION
**Pull request by caseyli92**
## DESCRIPTION OF PR:
* #106 
* #108 
* #109 

## Developer Checklist
- [x] Make code change
- [ ] Update tests
  - [ ] Update / create new tests
  - [ ] Ensure these tests have the expected behaviour
  - [ ] Test locally and ensure tests are passing
- [ ] Update documentation
  - [ ] Readme
  - [x] Docstrings
  - [x] Comments
  - [ ] Wiki

## Reviewer Checklist
- [ ] Check new code for code smells
- [ ] Check new tests
  - [ ] Ensure adequate coverage
  - [ ] Check for code smells within tests
- [ ] Check if documentation needs updating
  - [ ] Readme
  - [ ] Docstrings
  - [ ] Comments
  - [ ] Wiki

### CODE REVIEW

**Review by LukeParky — APPROVED** [review]
Looks very good, lots of improvements in documentation. I spotted a few things that I wasn't sure about so let's go over any that you want to talk about tomorrow morning.

File: src/digitaltwin/get_data_using_geoapis.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        The cleaned vector data as a GeoDataFrame.
+    """
+    # Convert column names to lowercase
+    fetched_data.columns = fetched_data.columns.str.lower()
+    # Move 'geometry' column to the end
+    fetched_data['geometry'] = fetched_data.pop('geometry')
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/digitaltwin/get_data_using_geoapis.py:47]: done

File: src/digitaltwin/setup_environment.py [hunk 2]
```diff
@@ -1,58 +1,93 @@
...
+    ------
+    ConnectionAbortedError
+        If the connection to the database fails.
+    """
+    try:
+        engine = get_connection_from_profile()
+        log.debug("Connected to PostgreSQL database successfully!")
+        return engine
+    except OperationalError as e:
+        raise ConnectionAbortedError("Connection to database failed. Please check .env file.") from e
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/digitaltwin/setup_environment.py:41]: This try and except block is newly added. I think previously you were raising 'ConnectionAbortedError', so I thought to keep it as it is. But sure, I will change it to what I have in the try-except block :)

File: src/digitaltwin/tables.py [hunk 3]
```diff
@@ -1,100 +1,201 @@
...
             session.rollback()
-
-
-def table_exists(engine, name, schema="public"):
-    """Check whether table already exist in the database"""
-    check_exists = sqlalchemy.inspect(engine).has_table(name, schema)
-    return check_exists
+            raise error
+        finally:
+            session.close()
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/digitaltwin/tables.py:201]: good spotting

File: src/digitaltwin/tables.py [hunk 4]
```diff
@@ -1,100 +1,201 @@
...
+    -------
+    None
+        This function does not return any value.
+
+    Raises
+    ------
+    Exception
+        If an error occurs during the execution of the query.
+    """
+    with Session(engine) as session:
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/digitaltwin/tables.py:193]: I removed the 'finally' block containing session.close(), added session.begin(), and kept the rest unchanged. Overall, there are some differences compared to what's in that document. Let me know what you think.

File: src/dynamic_boundary_conditions/hirds_rainfall_data_from_db.py [hunk 5]
```diff
@@ -51,104 +48,108 @@ def get_each_site_rainfall_data(
...
+        A DataFrame containing the rainfall data for the sites within the catchment area.
     """
+    # Get the site IDs within the catchment area
     sites_id_in_catchment = hirds_rainfall_data_to_db.get_sites_id_in_catchment(sites_in_catchment)
-
+    # Initialize an empty DataFrame to store the rainfall data
     rain_data_in_catchment = pd.DataFrame()
+    # Iterate over each site ID in the catchment area
     for site_id in sites_id_in_catchment:
+        # Retrieve the rainfall data for each site
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/dynamic_boundary_conditions/hirds_rainfall_data_from_db.py:151]: done :)

File: src/lidar/dem_metadata_in_db.py [hunk 6]
```diff
@@ -1,83 +1,200 @@
...
-    row = engine.execute(query)
-    return row.fetchone()[0]
+Base = declarative_base()
 
 
-def generate_dem(instructions):
+def get_hydro_dem_metadata(
+        instructions: Dict[str, Any],
+        catchment_boundary: gpd.GeoDataFrame) -> Tuple[str, str, str]:
+    """Get the hydrological DEM metadat~a."""
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/lidar/dem_metadata_in_db.py:32]: good spotting :)

File: src/dynamic_boundary_conditions/main_tide_slr.py [hunk 7]
```diff
@@ -23,91 +24,57 @@
...
+        BG-Flood model directory containing the uniform boundary input files.
 
-def remove_existing_boundary_input(bg_flood_dir: pathlib.Path):
-    # iterate through all files in the directory
+    Returns
+    -------
+    None
+        This function does not return any value.
+    """
+    # Iterate through all files in the directory
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/dynamic_boundary_conditions/main_tide_slr.py:43]: done :)

File: src/dynamic_boundary_conditions/main_tide_slr.py [hunk 8]
```diff
@@ -23,91 +24,57 @@
...
     for file_path in bg_flood_dir.glob('*_bnd.txt'):
-        # remove the file
+        # Remove the file
         file_path.unlink()
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/dynamic_boundary_conditions/main_tide_slr.py:46]: done

File: src/dynamic_boundary_conditions/hirds_rainfall_data_from_db.py [hunk 9]
```diff
@@ -1,46 +1,43 @@
...
+    pd.DataFrame
+        Filtered rainfall data for the requested duration.
     """
     if duration != "all":
+        # Filter the rain_data DataFrame to include only the relevant columns for the requested duration
         rain_data = rain_data[["site_id", "category", "rcp", "time_period", "ari", "aep", duration]]
     return rain_data
 
 
 def get_each_site_rainfall_data(
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/dynamic_boundary_conditions/hirds_rainfall_data_from_db.py:39]: changed to `get_one_site_rainfall_data`

File: src/digitaltwin/utils.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        The GeoDataFrame representing the catchment area.
+    to_crs : int, optional
+        Coordinate Reference System (CRS) code to convert the catchment area to. Default is 2193.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        The catchment area GeoDataFrame with the transformed CRS.
+    """
+    return catchment_area.to_crs(to_crs)
```
[the remark this answers is not in the document]
Comment (caseyli92) [reply src/digitaltwin/utils.py:126]: discussed already. leave it as it is.

**Pull request merged**